pkg/provider/cloud/openstack/provider.go

```diff
@@ -195,233 +199,377 @@ func validateExistingSubnetOverlap(networkID string, netClient *gophercloud.Serv
+		kubernetes.AddFinalizer(cluster, RouterCleanupFinalizer)
+		cluster.Spec.Cloud.Openstack.RouterID = router.ID
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to update RouterID in the cluster spec: %w", err)
 	}
 
-	// We only attach the router to the subnet if CloudProviderInfrastructure
-	// health status is not up, meaning that there was no successful
-	// reconciliations so far. This is to avoid hitting OpenStack API at each
-	// iteration.
-	// TODO: this is terrible, find a better way.
-	if cluster.Status.ExtendedHealth.CloudProviderInfrastructure != kubermaticv1.HealthStatusUp {
-		if kubernetes.HasFinalizer(cluster, RouterSubnetLinkCleanupFinalizer) {
-			if _, err = attachSubnetToRouter(netClient, cluster.Spec.Cloud.Openstack.SubnetID, cluster.Spec.Cloud.Openstack.RouterID); err != nil {
-				return nil, fmt.Errorf("failed to attach subnet to router: %w", err)
-			}
+	// Attach the new router to subnets
+	err = attachSubnetsIfNeeded(netClient, cluster)
+
+	return cluster, err
+}
+
+func attachSubnetsIfNeeded(netClient *gophercloud.ServiceClient, cluster *kubermaticv1.Cluster) error {
+	// Check if the router is already attached to the IPv4 subnet and attach to missing subnet if necessary
+	if cluster.Spec.Cloud.Openstack.SubnetID != "" {
+		err := linkSubnetToRouter(netClient, cluster, cluster.Spec.Cloud.Openstack.SubnetID, cluster.Spec.Cloud.Openstack.RouterID, RouterSubnetLinkCleanupFinalizer)
+		if err != nil {
+			return err
 		}
-		if kubernetes.HasFinalizer(cluster, RouterIPv6SubnetLinkCleanupFinalizer) {
-			if _, err = attachSubnetToRouter(netClient, cluster.Spec.Cloud.Openstack.IPv6SubnetID, cluster.Spec.Cloud.Openstack.RouterID); err != nil {
-				return nil, fmt.Errorf("failed to attach subnet to router: %w", err)
-			}
+	}
+
+	// Check if the router is already attached to the IPv6 subnetand attach to missing subnet if necessary
+	if cluster.Spec.Cloud.Openstack.IPv6SubnetID != "" {
+		err := linkSubnetToRouter(netClient, cluster, cluster.Spec.Cloud.Openstack.IPv6SubnetID, cluster.Spec.Cloud.Openstack.RouterID, RouterIPv6SubnetLinkCleanupFinalizer)
+		if err != nil {
+			return err
 		}
 	}
 
-	return cluster, nil
+	return nil
 }
 
-// TODO: Hey, you! Yes, you! Why don't you implement reconciling for Openstack? Would be really cool :)
-// func (os *Provider) ReconcileCluster(cluster *kubermaticv1.Cluster, update provider.ClusterUpdater) (*kubermaticv1.Cluster, error) {
-// 	return cluster, nil
-// }
+func linkSubnetToRouter(netClient *gophercloud.ServiceClient, cluster *kubermaticv1.Cluster, subnetID string, routerID string, finalizer string) error {
+	var ipAttached bool
+	router, err := getRouterIDForSubnet(netClient, subnetID)
+	ipAttached = err == nil && router != ""
+	if !ipAttached {
+		_, err := attachSubnetToRouter(netClient, subnetID, routerID)
+		kubernetes.AddFinalizer(cluster, finalizer)
+		if err != nil {
+			return err
+		}
```

Comment:
Is there a reason to add the finalizer even if the attach call failed? You're not immediately updating the Cluster object in etcd, but you're changing the cluster variable nontheless. Probably harmless, but maybe the AddFinalizer() should be done after the if?